### DISCUSSION TITLE
Sentinel 2 Level-2 Parallel Processing & Cloud Cover

### DISCUSSION THREAD

**Discussion by cshatto** [davidfrantz/force#268] [Q&A]
Hi all, 

I'm struggling to process Sentinel 2 imagery to level-2 and after having scoured the discussions here, I'm at my wits end. I have two general problems that could be related but I am unsure. Firstly, I was able to successfully process 10 Sentinel images and interpolate an NDVI time series. From here, I attempted to use the same parameter file (32 nproc / 2 nthread; which I realize now is overloading my system .. I'm running 40 CPUs with ~128RAM) to process the full year 2020 but it failed to process any image. Often it will kill processes (/usr/local/bin/force-l2ps_: line 161: 221824 Killed $EXE $TODO $PRM >> $LOGFILE) or save a logfile with 100% cloud cover (I've confirmed the DEM data type as mentioned in another discussion). 

At the moment, I am attempting to process tile T32TPU for the year 2020 but will later download the full time range. I've adjusted the nproc and nthread to fit my system a few different ways: 40/1, 32/1, 20/2, 16/2, 8/2. When using 20/2 and below, FORCE doesn't manage to start process images even after letting it run over the weekend. I know my system is capable since we've run FORCE before (see: https://github.com/davidfrantz/force/discussions/118). 

I hope I've laid out the parameters of my problem clearly, any help on this matter would be appreciated. 

Thanks in advance, 
Chris

**Comment by davidfrantz** [comment]
Hi @cshatto,

yes, it seems that you ran too many processes with respect to available RAM. With 8GB/process, you should be able to make it work. The ``killed`` message is happening as a direct consequence of this. Please see here for an illustration:

https://force-eo.readthedocs.io/en/latest/howto/l2-ard.html#parallel-processing

On your system, it should work with 16- processes (provided no other programs are interfering). 

Did you
- look into the logfiles for additional error messages?
- reset the queue after the unsuccessful attempts (i.e., changed DONE -> QUEUED)?
- check for "leftover" files that can prevent further processing? (*.lock files might be present if the process was killed during a write operation. If so, delete these manually, along with the corresponding files that were locked)

Cheers,
David

**Reply by cshatto** [reply]
Hi David, 

Thank you for your prompt response. I am familiar with the visual you provided but the issue is finding the proper balance within my system to utilize the "good" setup. I've attempted to run it on 16/2 before, along with 8/2 but after leaving it to run over the weekend, it still shows an ETA of a couple million seconds. During this process, it will skip the images with high cloud cover (showing 0% dc and 100% cc on the log files). It does this rather quickly, after about a minute. Other times, it will provide an AOD on the log file and say that its successfully written 0 items but still show 100% cc. Sometimes it says that there is say 34% dc but still 100% cc. I haven't come across any other errors in the log file when attempting to process Sentinel imagery. At the moment, I am running it on 16/1.

In response to the other questions: 
- I manually reset the queue after every unsuccessful attempt.
- I clear out all the .temp files but otherwise I have not seen any "leftover" files nor have I come across any .lock files. Where would those be stored? 


Thanks again for taking the time to help resolve this issue.
Chris

**Comment by davidfrantz** [comment]
Hi Chris,

skipping high cloud coverage images is the default and intended behavior. You can switch this off by setting the MAX_CLOUD_* parameters to 100.

The million seconds are of concern, though. I suspect that this is due to locked files. The presence of a lockfile will send the process to sleep until a timeout is reached. This could explain the unreasonably long times.

The lock files are stored in the Level 2 output folder. They are named like the other output files, but with a .lock extension.
Try doing ``find DIR -name '*.lock'`` (replace DIR with your output folder).

Cheers,
David

**Reply by cshatto** [reply]
Hi David, 

I figured as much with the skipped files, thank you for the confirmation. 

I don't have any .lock files in the output folder, or any files for that matter. Only the default 'europe' folder in which I have the datacube-definition.prj.


Chris

**Reply by davidfrantz** [reply]
hmm... can you attach the parameter file?

**Reply by cshatto** [reply]
Sure thing. Its saved as .prm but I changed it to .txt to upload here. 

[LEVEL2-skeleton.txt](https://github.com/davidfrantz/force/files/10662997/LEVEL2-skeleton.txt)

**Reply by davidfrantz** [reply]
Unfortunately, I do not see any obvious mistake....

Is there enough free space on the disc?

I would start debugging this by making the problem simpler, and then re-adding complexity. I suggest to process one of those images directly with force-l2ps and see what happens. I would also disable as many processing steps as possible (no tiling, no reprojection, no atmospheric correction, no resolution merging. no DEM, etc.). It is probably also best to use a clean output directory until we know what is going on.

**Reply by cshatto** [reply]
Hi David, 

Thanks for looking it over and for the recommendation. I will let you know here when I make any progress. 

Until then,
Chris

**Reply by cshatto** [reply]
Hi David, 

Thanks again for the advice. After starting simply with force-l2ps, the issue was that the srtm.tif had the wrong file extensions for the folder I was working in, which rendered the srtm.vrt useless. This occurred when I reorganized my files in preparation for the big download. I updated the file paths and built the .vrt again and it seems to have worked. Having changed this, I went ahead and added all the other processes and its successfully processing level2 images (force-level2) at an average rate of 35s for a little over 200 images (16nproc/2nthread). 

I suppose this is an underwhelming solution to the problem but I'm grateful for your help!

All the best, 
Chris
